src/EditorFeatures/Core/Implementation/IntelliSense/EditorCompletionImplementation/EditorAsyncCompletionItemManager.cs

```diff
@@ -0,0 +1,620 @@
+            // items based on a number that exists in the name of the item. However, when
+            // we are not after a dot (i.e. we're being brought up after <space> is typed)
+            // then we don't want to filter things. Consider the user writing:
+            //
+            //      dim i =<space>
+            //
+            // We'll bring up the completion list here (as VB has completion on <space>).
+            // If the user then types '3', we don't want to match against Int32.
+            if (filterText.Length > 0 && char.IsNumber(filterText[0]))
+            {
+                if (!IsAfterDot(data.Snapshot, session.ApplicableToSpan))
+                {
+                    session.Dismiss();
+                    return new EditorCompletion.FilteredCompletionModel(ImmutableArray<EditorCompletion.CompletionItemWithHighlight>.Empty, 0);
+                }
+            }
+
+            // We need to filter if a non-empty strict subset of filters are selected
+            var selectedFilters = data.SelectedFilters.Where(f => f.IsSelected).Select(f => f.Filter).ToImmutableArray();
+            var needToFilter = selectedFilters.Length > 0 && selectedFilters.Length < data.SelectedFilters.Length;
+
+            var initialListOfItemsToBeIncluded = new List<FilterResult>();
+            foreach (var item in data.InitialSortedList)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+
+                if (needToFilter && ShouldBeFilteredOutOfCompletionList(item, selectedFilters))
+                {
+                    continue;
+                }
+
+                if (MatchesFilterText(item.FilterText, item.DisplayText, GetMatchPriority(item), filterText, data.Trigger.Reason))
+                {
+                    initialListOfItemsToBeIncluded.Add(new FilterResult(item, filterText, matchedFilterText: true));
+                }
+                else
+                {
+                    if (data.Trigger.Reason == EditorCompletion.CompletionTriggerReason.Deletion ||
+                        data.Trigger.Reason == EditorCompletion.CompletionTriggerReason.Backspace ||
+                        data.Trigger.Reason == EditorCompletion.CompletionTriggerReason.Invoke ||
+                        filterText.Length <= 1)
+                    {
+                        initialListOfItemsToBeIncluded.Add(new FilterResult(item, filterText, matchedFilterText: false));
+                    }
+                }
+            }
+
+            var snapshotForDocument = data.InitialSortedList
+                .FirstOrDefault(i => i.Properties.ContainsProperty(CompletionItemSource.TriggerBuffer))?.
+                Properties.GetProperty<ITextBuffer>(CompletionItemSource.TriggerBuffer).CurrentSnapshot ?? data.Snapshot;
+
+            if (initialListOfItemsToBeIncluded.Count == 0)
+            {
+                var document = snapshotForDocument.GetOpenDocumentInCurrentContextWithChanges();
```

Comment:
We seem to use this snapshotForDocument a lot of places to fetch the document. Can we just do it once and get it over with?